### DISCUSSION TITLE
Systems and feature backlog

### DISCUSSION THREAD

**Discussion by FamroFexl** [FamroFexl/Circumnavigate#16] [General]
#### Currently, many of the main systems have been completed. 

There are several major systems yet to even be explored. These major systems are arguably the most complex and difficult to wrap your head around. (ahem, Wrapped Entity Path Finding and Wrapped Chunk Generation)

Main Systems:
- [x] Basic Debugging Tools
- [x] Vanilla Client Support
- [x] Integrated Client Support
- [x] Dedicated Server Support
- [x] LAN Server Support
- [x] Wrapped Chunk Loading
- [x] Wrapped Light Updates
- [x] Wrapped Entity Locations
- [x] Wrapped Entity Hitboxes
- [x] Wrapped Block Interaction
- [x] Wrapped Block Updates
- [x] No Over-Bounds Chunk Generation and Storage
- [x] No Hanging World Generation (stuck generating, never loads).
- [x] Proper Player Initial Spawning and Respawning
- [x] Proper Entity Spawning
- [x] Wrapped Nether Portals
- [x] Better World Creation GUI
- [x] Wrapped Nether
- [x] Wrapped Block Ticking
- [x] Wrapped Entity Path Finding
- [ ] Wrapped Structure Generation
- [ ] Wrapped Feature Generation
- [x] Wrapped Chunk Generation

Other systems:
- [ ] Chunk Shifting
- [x] Dynamic Curvature Shader
- [ ] Wrapped Commands
- [ ] Wrapped Datapacks
- [ ] Warp Teleportation at Edge of Client World (largely unnecessary, but possible)

**Comment by Skidamek** [comment]
Lan servers works perfectly fine, I am curious what about normal severs? There isn't yet any config. I guess currently you can configure this mod only when creating single player worlds. I can implement it if that's ok

And what do you mean by `Wrapped Entity Hitboxes`? I tested pve and pvp across borders and it works

**Reply by Skidamek** [reply]
And check your discord dm's please 🥺

**Reply by FamroFexl** [reply]
> Lan servers works perfectly fine, I am curious what about normal severs? There isn't yet any config. I guess currently you can configure this mod only when creating single player worlds. I can implement it if that's ok
> 
> And what do you mean by `Wrapped Entity Hitboxes`? I tested pve and pvp across borders and it works

Regular vanilla still requires you to create most worlds with custom features in the client, so that shouldn't be a feature the server needs to have. So far, there doesn't need to be any config, and I'd preferably like to keep it that way.

Wrapped Entity hitboxes is basically a function I implemented in the `WorldTransformer;splitAcrossBounds` method, which basically splits an entities' hitbox across the bounds it is occupying so it can calculate its collisions on all sides. I have yet to see if it works, and based on one issue of a pig suffocating in a border, it might not (although that could be pathfinding as well).

**Reply by Skidamek** [reply]
>Regular vanilla still requires you to create most worlds with custom features in the client, so that shouldn't be a feature the server needs to have. So far, there doesn't need to be any config, and I'd preferably like to keep it that way.

But then you can't just place this mod on server and use it... you first would need to create singleplayer world and upload it to server to use it

**Reply by FamroFexl** [reply]
> But then you can just place this mod on server and use it... you first would need to create singleplayer world and upload to to server to use it

True, but that doesn't majorly differ from vanilla minecraft, which just auto-generates a random world for you to start in. I have been using a server to test vanilla clients.

**Reply by Skidamek** [reply]
I disagree, this is that kind of mod which definitely should have possibility of configure server side. i can implement config which will only generate/load/work on dedicated servers so that on client you would only have your gui. 

The argument that mojang doesn't care doesn't make sense, they don't make any money out of it so they make gui to use on realms from where they do make money, but fundamental settings are still in server.properties. So since to use this mod you need to configure it beforehand, I believe that if someone installs this mod on their server it means they care and therefore having a config is fundamental

I don't know about you but personally i have never used those custom features on server anyway.

**Reply by FamroFexl** [reply]
> I disagree, this is that kind of mod which definitely should have possibility of configure server side. i can implement config which will only generate/load/work on dedicated servers so that on client you would only have your gui.
> 
> The argument that mojang doesn't care doesn't make sense, they don't make any money out of it so they make gui to use on realms from where they do make money, but fundamental settings are still in server.properties. So since to use this mod you need to configure it beforehand, I believe that if someone installs this mod on their server it means they care and therefore having a config is fundamental
> 
> I don't know about you but personally i have never used those custom features on server anyway.

I just don't know where you'd put config options. If you wanted to generate a few worlds, the inconvenience of using the client gui isn't that bad, but if you wanted to generate many worlds, or periodically refresh worlds, a config describing generation settings for a single world wouldn't be that useful. You'd have to implement it as a command. World bounds shouldn't be changed after the fact, so if the server.properties was to contain those values, they would only serve as a template for new worlds.

**Reply by FamroFexl** [reply]
And these additions are best done on an as-needed basis. If somebody is inconvenienced enough to put the effort in and open and issue, then it is worth considering.

**Reply by Skidamek** [reply]
> And these additions are best done on an as-needed basis. If somebody is inconvenienced enough to put the effort in and open and issue, then it is worth considering.

Done 😄

**Reply by Skidamek** [reply]
> I just don't know where you'd put config options. If you wanted to generate a few worlds, the inconvenience of using the client gui isn't that bad, but if you wanted to generate many worlds, or periodically refresh worlds, a config describing generation settings for a single world wouldn't be that useful. You'd have to implement it as a command. World bounds shouldn't be changed after the fact, so if the server.properties was to contain those values, they would only serve as a template for new worlds.

True but you likely already have desired size of your world no matter a seed, however also possibility of changing world size in the future would be great. (I know it would likely broke already built structures on the borders in the world and would require regeneration of new chunks but still)

**Comment by FamroFexl** [comment]
In regards to the **Wrapped Nether** system, there are several solutions:

1. `nether = overworld/8`: This would make all portals sync correctly without changes, but would basically turn off the nether in overworld's smaller than 64x64, because a dimension needs a minimum of an 8x8 area to exist in a wrapped fashion. The fog would be quite annoying at those small distances.
2. `nether = overworld + etc`: Nether portals reroute their coordinates (in some yet unspecified way). This would mean the size of the nether is unlinked from the size of the overworld, and portals just translate their coordinates to compensate. Coordinates could wrap around, so once you are past overworld coordinates in the nether, it just sends you to the other side.
4. `nether = overworld`: The nether is the same size as the overworld, meaning portals wouldn't scale down their coordinates, and a coordinate in the nether is the same coordinate in the overworld. This would remove the `nether = overworld/8` default, which is useful for fast travel.

**Reply by Skidamek** [reply]
Lets make `nether = overworld/x` where x is configurable with x = 8 by default

**Comment by FamroFexl** [comment]
Lots of blocks and entities will probably need specific transformations applied to each of them in order to work properly. I am thinking we focus on getting all the major systems above figured out before getting into the nitty-gritty of each specific block or entity. Many of the transformations are quite obvious to implement anyway. They aren't a source of problems, just a time sink. 

(I do wish that mixins had the capability to target every overrided member of a parent class at once and apply general transformations. Then, we could just selectively exclude child classes from being targeted. If most sub-classes need the transformation and only a few can't use it, it is much less time-consuming to include all and exclude some than exclude all and include most.)

**Reply by Skidamek** [reply]
it may have, just mcdev may not give it to you, worth asking on fabric cord or llama's lounge (man behind mixin extras - probably knows mixins best)

**Reply by Skidamek** [reply]
or we can try to generate mixins in runtime xD
that's definitely possible i believe that's how [connector](https://modrinth.com/mod/connector) works
